db/migrate/20170304172450_create_player_selections.rb

```diff
@@ -0,0 +1,11 @@
+class CreatePlayerSelections < ActiveRecord::Migration[5.0]
+  def change
+    create_table :player_selections do |t|
+      t.references :player, null: false
+      t.references :hero, null: false
+      t.string :role
+      t.references :composition, null: false
+      t.timestamps
+    end
```

Comment:
TLDR; The suggestion is to remove `player` and `hero` from this relation and use the (possibly) renamed join relation `player_heroes`.

I'm a little concerned here that we'd be denormalizing data contained in the `player_heroes` join relation by storing `player` and `hero` apart from each other in this relation. An alternative is to use the `player_heroes` join relation within this join relation, with the name resulting in `composition_player_heroes`. This has an important implication for the business logic rules for the app, however, in that a player creating a comp could only add `player_heroes` records to the comp. I think this helps us accomplish two things: 

1.  Prevents someone from creating a comp that includes a hero for a player that the player has no experience with. The idea there is that we can use our data model to prevent "invalid" combinations of data from occurring.

2. Using the `player_heroes` join relation in this join relation provides a concrete list of options for a player creating a comp. This has important fallout in at least a couple of ways I can think of:

  * It also makes it possible, given a list of players we want to make a comp for, to reduce that concrete list according to the hero role. Say I want to select two supports for my comp, given 6 players. In the UI, if I have a way to filter those 6 players according to support heroes, then I can be sure to pick a support for a player of interest without worrying if that player can fulfill the role I intend. 
  * Prevents people from making comps based on assumptions about what other players are comfortable playing. Using the `player_heroes` join relation allows individual players to signal what heroes they are comfortable and confident in playing for the sake of this tool (which I take to be more serious than quick play).

Does this make sense, and does it seem sensible to pursue?